src/encoder.js

```diff
@@ -270,41 +271,100 @@ export default class Encoder extends Extendable {
     });
   }
 
-  replaceExtensionPlaceholders(target: SuperPackedValueWithPlaceholders) : SuperPackedValue {
-    let enabledExtensions = Object.keys(this.extensions).map(e => +e).sort();
-    while (enabledExtensions.length > 0) {
-      let extensionPoint = enabledExtensions.pop();
-      let extension = this.extensions[extensionPoint];
-      let applyRecursively = extension.shouldApplyRecursively != null && extension.shouldApplyRecursively();
-
-      if (extensionPoint in this.placeholderMap) {
-        let placeholders = this.placeholderMap[extensionPoint];
-        for (let placeholderIndex = 0; placeholderIndex < placeholders.length; ++placeholderIndex) {
-          let placeholder = placeholders[placeholderIndex];
-          let extTarget = [];
-          if (extension.shouldSerialise == null || extension.shouldSerialise(placeholder.value)) {
-            if (extensionPoint < 8) {
-              extTarget.push(tags.EXTENSION3_BASE | extensionPoint);
-            } else {
-              extTarget.push(tags.EXTENSION_);
-              encodeUInt(extensionPoint, extTarget);
-            }
-            let furtherEnabledExtensions = applyRecursively
+  replaceExtensionPlaceholders(target: SuperPackedValueWithPlaceholders, extensions: Array<number>, memo: boolean = true) : SuperPackedValue {
+
+    // repeatedly apply all extensions until there is no more work to be done (# of iterations = depth of extension modified extension data)
+    while (Object.keys(this.placeholderMap).length > 0) {
+      let placeholderMap = this.placeholderMap;
+      let replacements = [];
+      let enabledExtensions = extensions.slice(0);
+
+      // run the logic for all extensions
+      while (enabledExtensions.length > 0) {
+        let extensionPoint = enabledExtensions.pop();
+        let extension = this.extensions[extensionPoint];
+        let applyRecursively = extension.shouldApplyRecursively != null && extension.shouldApplyRecursively();
+
+        if (extensionPoint in placeholderMap) {
+          let placeholders = placeholderMap[extensionPoint];
+          let originalSize = placeholders.length;
```

Comment:
Does `placeholders` change during the execution of the loop? If not, why store this?